pumpkin/src/server/json_config.rs

```diff
@@ -0,0 +1,92 @@
+use std::{fs, path::Path, sync::LazyLock};
+
+use log::warn;
+use pumpkin_config::op;
+use serde::{Deserialize, Serialize};
+
+pub static OPERATOR_CONFIG: LazyLock<tokio::sync::RwLock<OperatorConfig>> =
+    LazyLock::new(|| tokio::sync::RwLock::new(OperatorConfig::load()));
+
+#[derive(Deserialize, Serialize, Default)]
+#[serde(transparent)]
+pub struct OperatorConfig {
+    pub ops: Vec<op::Op>,
+}
+
+pub trait LoadJSONConfiguration {
+    #[must_use]
+    fn load() -> Self
+    where
+        Self: Sized + Default + Serialize + for<'de> Deserialize<'de>,
+    {
+        let path = Self::get_path();
+
+        let config = if path.exists() {
+            let file_content = fs::read_to_string(path)
+                .unwrap_or_else(|_| panic!("Couldn't read configuration file at {path:?}"));
+
+            serde_json::from_str(&file_content).unwrap_or_else(|err| {
+                panic!(
+                    "Couldn't parse config at {path:?}. Reason: {err}. This is probably caused by a config update. Just delete the old config and restart.",
+                )
+            })
+        } else {
+            let content = Self::default();
+
+            if let Err(err) = fs::write(path, serde_json::to_string_pretty(&content).unwrap()) {
+                eprintln!(
+                    "Couldn't write default config to {path:?}. Reason: {err}. This is probably caused by a config update. Just delete the old config and restart.", 
+                );
+            }
+
+            content
+        };
+
+        config.validate();
+        config
+    }
+
+    fn get_path() -> &'static Path;
+
+    fn validate(&self);
+}
+
+pub trait SaveJSONConfiguration: LoadJSONConfiguration {
+    // suppress clippy warning
+
+    fn save(&self)
+    where
+        Self: Sized + Default + Serialize + for<'de> Deserialize<'de>,
+    {
+        let path = <Self as LoadJSONConfiguration>::get_path();
+
+        let content = match serde_json::to_string_pretty(self) {
+            Ok(content) => content,
+            Err(err) => {
+                warn!(
+                    "Couldn't serialize operator config to {:?}. Reason: {}",
+                    path, err
+                );
+                return;
+            }
+        };
+
+        if let Err(err) = fs::write(path, content) {
+            warn!(
+                "Couldn't write operator config to {:?}. Reason: {}",
+                path, err
+            );
+        }
+    }
+}
+
+impl LoadJSONConfiguration for OperatorConfig {
+    fn get_path() -> &'static Path {
+        Path::new("ops.json")
```

Comment:
We should think about folder structure, maybe store it in ./data/ops.json

Reply:
I did not mean the data volume. (The volume should be renamed to pumpkin) All things that changed should be a separate folder like data and not inside the root folder.

The goal is to create a robust folder structure. Maybe I should make this an issue so it can be discussed on properly.

- pumpkin(.exe) (can be read and execute only)
- config (you could now make this folder red only)
  - configuration.toml
  - features.toml
- plugins (you could now make this folder read and execute only)
  - example(.dll)
  - example2(.dll)
  - example3(.dll)
- data (needs to be rw to work)
  - world
  - ops.json
  - (maby) perms.db

A folder structure like this is awesome if you use docker because now you can create volumes with restricting permissions.